app/validators/questionnaire_schema.py

```diff
@@ -0,0 +1,298 @@
+            "edit_block.question.answers",
+            "add_block.question.answers",
+            "remove_block.question.answers",
+            "edit_block.question_variants.answers",
+            "add_block.question_variants.answers",
+            "remove_block.question_variants.answers",
+        ]
+
+        for match in parse("$..id").find(self.schema):
+            full_path = str(match.full_path)
+            if not any(ignored_path in full_path for ignored_path in ignored_paths):
+                yield str(match.full_path)[:-3], match.value
+
+    @cached_property
+    def answer_id_to_option_values_map(self):
+        answer_id_to_option_values_map = defaultdict(set)
+
+        for answer in self.answers:
+            if "options" not in answer:
+                continue
+
+            answer_id = answer["id"]
+            option_values = [option["value"] for option in answer["options"]]
+
+            answer_id_to_option_values_map[answer_id].update(option_values)
+
+        return answer_id_to_option_values_map
+
+    @cached_property
+    def answers(self):
+        for question, _ in self.questions_with_context:
+            for answer in question["answers"]:
+                yield answer
+
+    @lru_cache
+    def get_answer(self, answer_id):
+        return self.answers_with_context[answer_id]["answer"]
+
+    @lru_cache
+    def get_group(self, group_id):
+        return self.groups_by_id[group_id]
+
+    @lru_cache
+    def get_section(self, section_id):
+        return self.sections_by_id[section_id]
+
+    @lru_cache
+    def get_block(self, block_id):
+        return self.blocks_by_id[block_id]
+
+    @lru_cache
+    def get_blocks(self, **filters):
+        conditions = []
+        for key, value in filters.items():
+            conditions.append(f'@.{key}=="{value}"')
+
+        if conditions:
+            final_condition = " & ".join(conditions)
+            return jp.match(f"$..blocks[?({final_condition})]", self.schema)
+        return self.blocks
+
+    @lru_cache
+    def get_block_key_context(self, block_id, key_name):
+        """
+        Get all dicts that contain `key_name` within a block
+        :param block_id: the id of the block to search
+        :param key_name: the key to find
+        :return: list of dicts containing the key name, otherwise returns None
+        """
+        matches = []
+        for match in parse(f"$..{key_name}").find(self.blocks_by_id[block_id]):
+            matches.append(match.context.value)
+        return matches
+
+    @lru_cache
+    def get_other_blocks(self, block_id_to_filter, **filters):
+        conditions = []
+        for key, value in filters.items():
+            conditions.append(f'@.{key}=="{value}"')
+
+        if conditions:
+            final_condition = " & ".join(conditions)
+            return jp.match(
+                f'$..blocks[?(@.id!="{block_id_to_filter}" & {final_condition})]',
+                self.schema,
+            )
+        return self.blocks
```

Comment:
This has been made generic, but only seems to be used for one purpose. It also seems unnecessary to write a JSONPath query builder when the calling method could just construct the required JSONPath. I also think that if we're going to write generic methods there should be independent tests for them.